[Task] Add and enroll habits feature has been added

luchward opened the pull request:
**Restrictions:**

- refactoring of calendar components is out of scope;
- writing tests is out of scope.

**Results:**

- user story [#1000](https://github.com/ita-social-projects/GreenCity/issues/1000) has been completed;
- components that display the profile page have been refactored to meet mockup requirements;
- profile-dashboard component displays habits in progress and acquired;
- profile-dashboard component lets user to enroll and unenroll his/her/its habits;
- hyperlinks to user's profiles on social networks are displayed on the profile page;
- habits are displayed in corresponding localization;
- habit-service and habit-assign-service have been created; thay implement all methods of corresponding backend controllers.

![image](https://user-images.githubusercontent.com/58164899/103431535-1c9df600-4bda-11eb-81b6-a77a4b9eb33f.png)

JimHolman1199 reviewed (changes requested):
incorrect alignment of the whole page.
how it looks : 
<img width="718" alt="Screenshot 2020-12-30 at 23 44 56" src="https://user-images.githubusercontent.com/46226026/103382865-4af7d480-4af9-11eb-9207-53d4eb9fa9a7.png">
how it should be : 
<img width="1235" alt="Screenshot 2020-12-30 at 23 49 34" src="https://user-images.githubusercontent.com/46226026/103382989-c2c5ff00-4af9-11eb-97a4-33aa39344087.png">

use 'container' class! 

missing space
<img width="598" alt="Screenshot 2020-12-30 at 23 50 57" src="https://user-images.githubusercontent.com/46226026/103383050-f86ae800-4af9-11eb-8804-8e51864b4bc8.png">

JimHolman1199 commented on src/app/component/user/components/profile/profile-dashboard/profile-dashboard.component.ts:
```diff
@@ -1,50 +1,68 @@
+      .pipe(take(1))
+      .subscribe((response: Array<HabitAssignInterface>) => {
+        const sortedHabits = response
+          .sort((a, b) => (a.habit.id > b.habit.id) ? 1 : (b.habit.id > a.habit.id) ? -1 : 0);
```
Move this array "sort" operation to a separate statement.

luchward replied:
fixed

JimHolman1199 commented on src/app/service/habit/habit.service.ts:
```diff
@@ -0,0 +1,52 @@
+    return this.http.get<Array<string>>(`${habitLink}/tags?lang=${this.language}`);
+  }
+
+  getHabitsByTagAndLang(page: number, size: number, sort: string = 'asc', tags: Array<string>): Observable<HabitListInterface> {
```
Default parameters should be last.

luchward replied:
fixed

JimHolman1199 reviewed (commented):
consider writing unit tests because the coverage of the new code is very low

luchward commented:
JimHolman1199 Thank you for your attentiveness.

However there are not mockups drawn for 1200 px size:

![image](https://user-images.githubusercontent.com/58164899/103460963-67775500-4d23-11eb-810b-9990cba94c0e.png)

In this case I have implemented a sequence of container sizes:

- more or equal 320 px - fluid;
- more or equal 576 px - fluid;
- more or equal 768 px - 740 px;
- more or equal 1024 px - 960 px;
- more or equal 1440 px - 1140 px.

Using of container class leads to incorrect alignment:

![Безіменний](https://user-images.githubusercontent.com/58164899/103461552-d5be1680-4d27-11eb-90e2-6578b2393bdf.png)

and I don't wanna handle it, so I implemented my own way.

JimHolman1199 commented:
![Screenshot 2021-01-03 at 00 23 04](https://user-images.githubusercontent.com/46226026/103467680-251d3a80-4d5a-11eb-9bf4-c02c54a7db89.png)
![Screenshot 2021-01-03 at 00 24 38](https://user-images.githubusercontent.com/46226026/103467681-26e6fe00-4d5a-11eb-841d-06c0d22a2914.png)
calendar is not aligned to other content

You will need to use a predefined grid system, as your page will not match other pages.
![Screenshot 2021-01-03 at 00 26 24](https://user-images.githubusercontent.com/46226026/103467698-5269e880-4d5a-11eb-80af-bac5a281fa8b.png)

JimHolman1199 reviewed (changes requested):
fix page layout

JimHolman1199 commented on src/app/component/user/components/habit/all-habits/all-habits.component.ts:
```diff
@@ -1,10 +1,14 @@
+import { map } from 'rxjs/operators';
+import { HabitInterface, HabitListInterface } from '../../../../../interface/habit/habit.interface';
+import { singleNewsImages } from '../../../../../image-pathes/single-news-images';
+import {HabitAssignService} from '@global-service/habit-assign/habit-assign.service';
```
Remove this unused import of 'HabitAssignService'

luchward replied:
fixed

luchward commented:
I know about calendars, but I can not refactor every component I touch. It is infinite task.
What concerns to layout, as I understand some mockups don't meet styleguide, but it is not my resposibility. I follow mockup requirements for the particular page.

luchward commented:
I came up with a solution to change the last breakpoint from 1440 to 1200 px

JimHolman1199 reviewed (changes requested):
something went wrong
<img width="732" alt="Screenshot 2021-01-05 at 22 26 37" src="https://user-images.githubusercontent.com/46226026/103696464-124b7580-4fa7-11eb-8093-41145199b901.png">

mr-pavlenko commented on src/app/component/user/components/profile/profile-widget/profile-header/profile-header.component.scss:
```diff
@@ -1,153 +1,252 @@
+    }
+
+    .rate {
+      width: 73px;
```
Why is the width fixed?

luchward replied:
This block works as background for the text and has fixed width due to mockup requirements

mr-pavlenko commented on src/app/component/user/components/profile/profile-widget/profile-header/profile-header.component.scss:
```diff
@@ -1,153 +1,252 @@
+
+    .social {
+      display: flex;
+      width: 288px;
```
Why is the width fixed?

luchward replied:
Outer block has fixed width according to mockup and particular alignment of content, so this block has fixed width as well.

mr-pavlenko commented on src/app/service/habit/habit.service.ts:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  ngOnDestroy(): void {
+    this.destroy.next(null);
```
Why did you use null?

luchward replied:
my fault
